File: powerstrip-java/src/main/java/powerstrip/PowerStrip.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package powerstrip;

/**
 *
 * @author wmacevoy
 */
public class PowerStrip {
    
    public static final int MODE_ON = 0;
    
    private int outlets;
    private OutletState [] outletStates;
    private Mode mode;
    

    public PowerStrip(int _outlets) {
        if (_outlets <= 0) {
            throw new IllegalArgumentException("outlets (" + _outlets + ") must be positive.");
        }
        outlets=_outlets;
        outletStates = new OutletState[outlets];
        for (int outlet = 0; outlet < outlets; ++outlet) {
            outletStates[outlet] = OutletState.UNPOWERED;
        }
    }
    public int getOutlets() { return outlets; }   

    public void setMode(Mode _mode) {
        mode = _mode;
    }
    
    public Mode getMode() {
        return mode;
    }
    
    public void outletOk(int outlet) {
        if (outlet < 0 || outlet >= outlets) {
            throw new IllegalArgumentException("outlet (" + outlet + ") must be between 0 and " + (outlets-1) + ".");
        }
    }
    public void setOutletState(int outlet, OutletState state) {
        outletOk(outlet);
        outletStates[outlet]=state;
    }
    
    public OutletState getOutletState(int outlet) {
        outletOk(outlet);
        switch(mode) {
            case OFF: return OutletState.UNPOWERED;
            case ON: return OutletState.POWERED;
            case REMOTE: return outletStates[outlet];
        }
        throw new IllegalStateException();
    }
    
    public void setModeOff() { setMode(Mode.OFF); }
    public void setModeOn() { setMode(Mode.ON); }
    public void setModeRemote() { setMode(Mode.REMOTE); }

}

File: powerstrip-cpp/src/PowerStrip.cpp
#include "PowerStrip.h"
#include <stdexcept>
#include <sstream>
#include <iostream>

namespace powerstrip {
  PowerStrip::PowerStrip(int _outlets)
    : outlets(_outlets > 0 ? _outlets : 0), outletStates(outlets, OutletState::UNPOWERED), mode(Mode::REMOTE)
  {
    std::cout << "made PowerStrip@" << ((void*) this) << std::endl;
    
    if (outlets <= 0) {
      std::ostringstream oss;
      oss << "outlets (" << outlets << ") must be positive.";
      throw std::out_of_range(oss.str());
    }
  }

  int PowerStrip::getOutlets() const {
    return outlets;
  }

  void PowerStrip::setMode(Mode _mode) {
    mode = _mode;
  }

  Mode PowerStrip::getMode() const {
    return mode;
  }

  void PowerStrip::outletOk(int outlet) const {
    if (outlet < 0 || outlet >= outlets) {
      std::ostringstream oss;
      oss << "outlet (" << outlet << ") must be between 0 and " << (outlets-1) << ".";
      throw std::out_of_range(oss.str());
    }
  }

  void PowerStrip::setOutletState(int outlet, OutletState state) {
    outletOk(outlet);
    outletStates[outlet]=state;
  }

  OutletState PowerStrip::getOutletState(int outlet) const {
    outletOk(outlet);
    if (mode == Mode::ON) return OutletState::POWERED;
    else if (mode == Mode::OFF) return OutletState::UNPOWERED;
    else return outletStates[outlet];
  }

  void PowerStrip::setModeOff() { setMode(Mode::OFF); }
  void PowerStrip::setModeOn() { setMode(Mode::ON); }
  void PowerStrip::setModeRemote() { setMode(Mode::REMOTE); }

  PowerStrip::~PowerStrip() { std::cerr << "cleaned up PowerStrip" << std::endl; }
}
